src/ray/raylet/node_manager.cc

```diff
@@ -1843,6 +1844,120 @@ void NodeManager::HandleCancelResourceReserve(
+      std::string error_msg = absl::StrFormat(
+          "Cannot resize unit instance resource '%s'. Unit instance resources "
+          "(e.g., GPU) cannot be resized dynamically.",
+          resource_name);
+      send_reply_callback(Status::InvalidArgument(error_msg), nullptr, nullptr);
+      return;
+    }
+  }
+
+  // Get current local resources and convert to resource maps
+  const auto &current_resources =
+      cluster_resource_scheduler_.GetLocalResourceManager().GetLocalResources();
+  auto current_total_map =
+      current_resources.GetTotalResourceInstances().ToNodeResourceSet().GetResourceMap();
+  auto current_available_map = current_resources.GetAvailableResourceInstances()
+                                   .ToNodeResourceSet()
+                                   .GetResourceMap();
+
+  // Calculate delta resource map (target - current) and validate
+  absl::flat_hash_map<std::string, double> delta_resource_map;
+  for (const auto &[resource_name, target_value] : target_resource_map) {
+    double current_total = 0.0;
+    double current_available = 0.0;
+
+    if (auto total_it = current_total_map.find(resource_name);
+        total_it != current_total_map.end()) {
+      current_total = total_it->second;
+    }
+
+    if (auto available_it = current_available_map.find(resource_name);
+        available_it != current_available_map.end()) {
+      current_available = available_it->second;
+    }
+
+    double delta_value = target_value - current_total;
+
+    // Check if applying delta would make available resources negative
+    if (double new_available = current_available + delta_value; new_available < 0.0) {
+      std::string error_msg = absl::StrFormat(
+          "Cannot resize %s to %g: would make available resources negative "
+          "(current_available=%g, delta=%g, new_available=%g). ",
+          resource_name,
+          target_value,
+          current_available,
+          delta_value,
+          new_available);
+      send_reply_callback(
+          Status::RpcError(error_msg, grpc::StatusCode::FAILED_PRECONDITION),
+          nullptr,
+          nullptr);
+      return;
+    }
+
+    if (delta_value != 0.0) {
+      delta_resource_map[resource_name] = delta_value;
+    }
+  }
+
+  // Convert the delta resource map to NodeResourceInstanceSet and apply
+  if (!delta_resource_map.empty()) {
+    NodeResourceSet delta_resources(delta_resource_map);
+    NodeResourceInstanceSet delta_instances(delta_resources);
+
+    // Apply deltas for each resource
+    for (const auto &resource_id : delta_resources.ExplicitResourceIds()) {
+      const auto &instances = delta_instances.Get(resource_id);
+      cluster_resource_scheduler_.GetLocalResourceManager().AddLocalResourceInstances(
+          resource_id, instances);
+    }
+  }
+
+  // Get updated resource state and populate reply
+  const auto updated_resources =
+      cluster_resource_scheduler_.GetLocalResourceManager().GetLocalResources();
```

Comment:
```suggestion
  const auto &updated_resources =
      cluster_resource_scheduler_.GetLocalResourceManager().GetLocalResources();
```
I also looked at the definition and actually realized that GetLocalResources returns by value, can you update it to return by const ref